### PR TITLE
midproject commit

### PR DESCRIPTION
**Pull request by tonycolucci**
Submission for Mid-Project Review

### CODE REVIEW

**Comment by naf445** [comment]
testing what happens when I comment here (you may see a few of these)

**Review by naf445 — COMMENTED** [review]
testing what happens when I comment here (you may see a few of these) #2

File: README.md [hunk 1]
```diff
@@ -55,7 +56,7 @@ Create a clean, durable website with public access to the data
 ## Backlog
 Label | Content | Score 
 ------- |------| ----------- 
-Theme2.Epic2.Story1 | Create AWS Instancee upon which web app can be deployed | 1
```
Comment (naf445) [inline README.md]: testing what happens when I comment here (you may see a few of these)

**Comment by naf445** [comment]
> testing what happens when I comment here (you may see a few of these) #2

**Comment by naf445** [comment]
This pull request review is part of our MSiA AVC class requirements and on my side will consist of me beginning with the repository in it's current state attempting to recreate the data retrieval & storage steps as well as the setting up of both a local sqlite database and an AWS RDS database (data pipeline). After that I will commence with more general code review of other files in the project. 

I will first be giving updates here in the *Conversation* tab on my progress, problems, and suggestions on my initial attempt at recreating the data pipeline. Following that I will give explicit feedback on each of the following categories:

- Functionality
- Readability & code quality
- Overall design
- Testing
- Documentation


giphy

**Comment by naf445** [comment]
Ok then - my first step was to open the README for some guidance on setup. We have discussed this, but there are no setup instructions for the app or data pipeline process so I feel as if i'm sort of swimming through murky waters. Obviously some explicit instructions on what files and commands and environments and configurations to use and how to use them and all of that sort of stuff would help elucidate my path forward so I would **definitely** focus on beefing that up.

**Comment by naf445** [comment]
Alright i'm assuming before doing anything I should set up a virtual environment using the requirements.txt file so I have do that. Just for reference this is the list of packages which were installed when I used the requirements.txt for pip installation. 

|Package          |Version | |Package          |Version|
| ------------- | ------------- |-| ------------- | ------------- |
|atomicwrites  |   1.3.0   ||pip              |19.1.1  |
|attrs|            19.1.0  | |pluggy|           0.11.0| 
|awscli|           1.16.163||py               |1.8.0   |
|boto3|            1.9.153| |pyaml|            18.11.0| 
|botocore|         1.12.153||pyasn1           |0.4.5   |
|certifi |          2019.3.9||PyMySQL|          0.9.3|   
|chardet|          3.0.4|   |pytest           |4.0.1   |
|Click|            7.0|     |python-dateutil|  2.8.0|  
|colorama|         0.3.9|   |pytz             |2019.1  |
|docutils|         0.14|    |PyYAML|           5.1|   
|Flask            |1.0.2   ||requests         |2.22.0|  
|Flask-SQLAlchemy| 2.3.2|   |rsa|              3.4.2|   
|idna|             2.8     ||s3transfer       |0.2.0|   
|itsdangerous|     1.1.0|   |setuptools|       41.0.1|  
|Jinja2           |2.10.1  ||six              |1.12.0  |
|jmespath|         0.9.4|   |SQLAlchemy|       1.3.1|   
|MarkupSafe       |1.1.1  | |urllib3          |1.25.2  |
|more-itertools|   7.0.0|   |Werkzeug|         0.15.4|  
|numpy            |1.16.3  ||wheel            |0.33.4 |
|pandas|           0.24.2|

**Comment by naf445** [comment]
As I wade through the murky waters of your **documentation-less** repository and have set up my virtual environment, my next step is to use the `copydata_bash` file to try and retrieve the data in its raw form and transfer it to my own S3 bucket.

From examining the script it seems that I need to provide an S3 destination bucket as a parameter. Alright I then ran the command `bash copydata_bash s3://tonys-data` from the root directory.

First issue, AWS CLI is not installed on my new virtual environment (I don't think at least, I may be confused on this issue) so I would suggest adding this into the requirements (at the end I will give the updated list of requirements I have and see if it matches the initial list I put in the previous comment). I will go ahead and run `pip install --upgrade awscli` to get it installed and configured with my account.

After these steps the script seemed to run fine and the data appeared in my bucket on AWS.

**Comment by naf445** [comment]
Moving right along - my next step I believe is to set up an instance of a database with a schema attached. Ideally there would be some type of option whether I want to work with a local sqlite database file or work with RDS on AWS, it seems only to be configured for AWS RDS currently so I would take a look at developing that functionality when you get a chance. 

I see that `sqlalchemy_generate_table.py` looks for environmental variables previously configured so I will assume I need to edit the `.mysqlconfig` file. I will do so by putting in my information into the `.mysqlconfig` file. I then ran `source .mysqlconfig` and confirmed it worked successfully as `echo $MYSQL_USER` displayed my username. 

As a sidenote I noticed you included your endpoint/host when you committed & uploaded to GitHub. I don't think that is a security issue b/c I would assume you have to have permission to read/write from it anyways but it may be something worth thinking about and not including in your repo for others to see.  

Back to the action - OK now I that I have configured mysql with the aptly named .mysqlconfig I run `python sqlalchemy_generate_table.py`.

I get an error back saying `ModuleNotFoundError: No module named 'pymysql'` so it may not be in the requirements. I will `pip install pymysql` and try again.

Try #2: I received an error because I forgot to change my security group rules on my RDS instance to allow incoming traffic from my personal computer. After correcting this error, I re-ran the script and received this error:

```bash
(tony_virt_env) nathanfranklin$ python sqlalchemy_generate_table.py

Traceback (most recent call last):
  File "sqlalchemy_generate_table.py", line 46, in <module>
    throw1 = Track(id = 1, hand_rl = "R", throw_type = "F", stall = 2, completion = 1, Xdist = 11, Ydist = 3)  
NameError: name 'Track' is not defined
```

From looking through the file, I believe the error is that you define your class `Throw` in line 11 but down in line 46 where my error occurred you refer to this as class `Track`.

**Comment by naf445** [comment]
That concludes my initial attempt at recreating your data pipeline portion of the project, and it ended unsuccessfully as I couldn't instantiate my database instance. Though I think it is one or two minor changes away from working successfully and allowing me to finally reach the other side of the metaphorical murky lake i've been swimming through. 

Below I will begin the critique more explicitly along the 5 dimensions specified in our assignment.

File: sqlalchemy_generate_table.py [hunk 2]
```diff
@@ -37,20 +40,13 @@ def __repr__(self):
...
-track_record = session.query(Track.title, Track.album).filter_by(artist="Britney Spears").first()
-print(track_record)
-query = "SELECT * FROM tracks WHERE artist LIKE '%%Britney%%'"
-df = pd.read_sql(query, con=engine)
-print(df)
-session.close()
+# Session = sessionmaker(bind=engine)  
+# session = Session()
+# # add a record/track
+# throw1 = Track(id = 1, hand_rl = "R", throw_type = "F", stall = 2, completion = 1, Xdist = 11, Ydist = 3)
```
Comment (naf445) [inline sqlalchemy_generate_table.py]: I think you forgot to change `Track` to `Throw` here (line 46)

**Comment by naf445** [comment]
## Functionality
The `copydata_bash` file did work as planned. It may be a good idea to add some functionality of whether to store the files locally or to send to S3 or to do both.

The `sqlalchemy_generate_table.py` file did not run successfully but I believe this is due to a change which has been mentioned above.

## Readability & code quality
The bash script is simple enough, and the `sqlalchemy_generate_table.py` file is relatively straightforward also as it defines the schema, creates a DB connection, and then opens a session and commits a track to the DB we should have set up.

I think the organization is fine for now because it is only doing a few simple tasks but I think will have to be more thought out when more complex operations are necessary. 

The current variable/function/class names make sense and seem apt. Obviously, currently it simply sets up a database and adds a track, and in the future you will want to only set up the database if one doesn't already exist, and you will want to build functionality which adds new items to the database only if they need to be and turn that into a function or something similar which you could call multiple times with different entries. I'm sure there are many ways to go about doing this and depending on how you will be using the RDS instance.

## Overall design
The current codebase makes sense in the overall context of what the application needs to be at this point in its development. I think it lays a solid groundwork for a good workflow to keep developing the app and a good basis for the infrastructure which will underly the app.

It's hard to say how the current design will influence future development as I think the structure will change a lot, but one important thing to keep in mind would be allowing the user to specify options in how they want things to be set up and allow the app to work with any of those configurations if possible (storing data locally vs. S3 / using RDS vs local sqlite)

## Testing
No current testing implemented

## Documentation
- Minimal documentation in its current form. 
- No method/class documentation (although there isn't too much to document so it may even be worth waiting until the codebase is closer to it's final form to avoid spending lots of time documenting things which will be changed anyways).
- There is some commentary in the form of comments
- The README currently does not explain important points such as:
     + how someone else can get the code, build it, and run it, and get all functionality working without having to ask the developer questions
     + overall design / approach to the problem
- Commit messages are meaningful but lack detail and don't follow best practices (a severe crime which I am also guilty of). Below are the recommended guidelines for commit messages found [here](https://github.com/trein/dev-best-practices/wiki/Git-Commit-Best-Practices):
     + Begin your message with a short summary of your changes (up to 50 characters as a guideline). Separate it from the following body by including a blank line. The body of your message should provide detailed answers to the following questions: – What was the motivation for the change? – How does it differ from the previous implementation? Use the imperative, present tense («change», not «changed» or «changes») to be consistent with generated messages from commands like git merge.

File: sqlalchemy_generate_table.py (hunk 2 above)
Comment (tonycolucci) [reply sqlalchemy_generate_table.py]: Thanks for catching that. I've fixed that in my most recent update.

**Comment by tonycolucci** [comment]
I've added a significant amount of changes to the README to provide instructions for recreating the app.

**Pull request merged**